packages/rum-core/src/domain/contexts/pageStateHistory.ts

```diff
@@ -38,67 +63,54 @@ export function startPageStateHistory(): PageStateHistory {
       if (!event.isTrusted) {
         return
       }
+      const startTime = event.timeStamp as RelativeTime
 
       if (event.type === DOM_EVENT.FREEZE) {
-        addPageState(PageState.FROZEN)
+        addPageState(PageState.FROZEN, startTime)
       } else if (event.type === DOM_EVENT.PAGE_HIDE) {
-        addPageState((event as PageTransitionEvent).persisted ? PageState.FROZEN : PageState.TERMINATED)
+        addPageState((event as PageTransitionEvent).persisted ? PageState.FROZEN : PageState.TERMINATED, startTime)
       } else {
-        addPageState(getPageState())
+        addPageState(getPageState(), startTime)
       }
     },
     { capture: true }
   )
 
   return {
-    findAll(startTime: RelativeTime, duration: Duration) {
-      const entries: PageStateEntry[] = []
-      const endTime = addDuration(startTime, duration)
-      for (let i = pageStateEntries.length - 1; i >= 0; i--) {
-        const { startTime: stateStartTime } = pageStateEntries[i]
-
-        if (stateStartTime >= endTime) {
-          continue
-        }
-
-        entries.unshift(pageStateEntries[i])
-
-        if (stateStartTime < startTime) {
-          break
-        }
+    findAll: (startTime: RelativeTime, duration?: Duration): PageStateServerEntry[] | undefined => {
+      const pageStateEntries = pageStateHistory
+        .findAll(startTime, duration)
+        .reverse()
+        .map((pageState) => {
+          const correctedStartTime = startTime > pageState.startTime ? startTime : pageState.startTime
+          const recenteredStartTime = elapsed(startTime, correctedStartTime)
+
+          return {
+            state: pageState.state,
+            start: toServerDuration(recenteredStartTime),
+          }
+        })
+
+      if (pageStateEntries.length > 0) {
+        return pageStateEntries
       }
```

Comment:
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**💬 suggestion:** ‏a comment to explain what is the intent here would be helpful.

Reply:
Let me know if that clearer